File: adventOfCode/03/solution.py
import numpy as np
import re


with open("input.txt", "r") as inputfile:
    lines = [line.replace("\n", "") for line in inputfile.readlines()]


def parse_line(line):
    pattern = r"#([0-9]+) @ ([0-9]+),([0-9]+): ([0-9]+)x([0-9]+)"
    match = re.match(pattern, line)
    if not match:
        raise Exception(f"No pattern match: {line}")
    groups = match.groups()
    return dict(zip(
        ["id", "x", "y", "width", "height"],
        map(int, groups)
    ))


parsed_lines = list(map(parse_line, lines))
shape = (
    max(line["y"] + line["height"] + 1 for line in parsed_lines),
    max(line["x"] + line["width"] + 1 for line in parsed_lines)
)
fabric = np.zeros(shape, int)


# Part One


for line in parsed_lines:
    fabric[
        line["y"]:line["y"] + line["height"],
        line["x"]:line["x"] + line["width"]
    ] += 1

print(len(np.transpose((fabric > 1).nonzero())))


# Part Two


for line in parsed_lines:
    rect = fabric[
        line["y"]:line["y"] + line["height"],
        line["x"]:line["x"] + line["width"]
    ]
    if len(np.transpose((rect > 1).nonzero())) == 0:
        print(line["id"])

File: adventOfCode/03/solution.js
var fs = require('fs');

var lines = fs.readFileSync('input.txt', 'ascii').split('\n');

const lineRegex = /#([0-9]+) @ ([0-9]+),([0-9]+): ([0-9]+)x([0-9]+)/;

function parseLine(line) {
  const match = line.match(lineRegex);
  return {
    'id': parseInt(match[1]),
    'x': parseInt(match[2]),
    'y': parseInt(match[3]),
    'width': parseInt(match[4]),
    'height': parseInt(match[5])
  }
}

const parsedLines = lines.filter((x) => x).map(parseLine);

const width = Math.max(...parsedLines.map((line) => {
  return line.x + line.width;
})) + 1;
const height = Math.max(...parsedLines.map((line) => {
  return line.y + line.height;
})) + 1;

let fabric = new Array(height).fill(null);
fabric.forEach((x, i) => {
  fabric[i] = new Array(width).fill(0);
});

parsedLines.forEach((line) => {
  for (let j = line.y; j < line.y + line.height; j++) {
    for (let i = line.x; i < line.x + line.width; i++) {
      fabric[j][i] += 1;
    }
  }
});

const overlaps = fabric.map((row) => {
  return row.map((x) => x > 1).reduce((a, b) => a + b);
}).reduce((a, b) => a + b);

console.log(overlaps);
